[addons] remove not needed part and add new version check way

AlwinEsch opened the pull request:
First step of the big rework

## Description
This remove on first commit the never used GetSettings add-on function. This was thought to define addon settings by the code and not with xml, but nowhere used and becomes removed.

The other two commits add the first part of the new version way. This is thought to prevent e.g. the `GetVersion` function on PVR. There becomes everything added to one header (versions.h), reason to have in one is to read from them with cmake (on next changes in new coming request) and write them to a `addon.xml.in`.

## Motivation and Context
Big binary add-on system rework.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] remove never used functions

This removes function who was never used on all addon types to get
addon settings given by code and not with xml.

AlwinEsch committed:
[addons] add new addon versions.h header

Become used to have on addon code and also on build a place to identify
the versions. This prevent in future a lot of work on version changes and
a recompile is mostly enough.

MartijnKaijser commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -15232,7 +15232,13 @@ msgctxt "#29802"
+#. The label in the add-ons dll class to report about outdated add-on api version
+#: xbmc/addons/AddonDll.cpp
+msgctxt "#29803"
+msgid "Using incompatible API version for type '%s'.\nKodi API version = '%s', add-on API version '%s'\n\nPlease check for updates or contact developer of add-on."
```
Please do not use line breaks in strings if they are not really needed.

notspiff commented:
you want me to merge up front ?

AlwinEsch commented:
Want to create on all repo addons the request first, to reduce time for this changes.

tamland commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -461,5 +398,31 @@ ADDON_STATUS CAddonDll::TransferSettings()
+                                           kodi::addon::GetTypeName(type),
+                                           kodiVersion,
+                                           addonVersion);
+    CGUIDialogOK::ShowAndGetInput(CVariant{heading}, CVariant{text});
```
Don't make gui calls from deep down in the addons system please. We shouldn't have random functions adding popous. Instead this should be in the gui code tied to some user action (or notification if it's a background task)

AlwinEsch replied:
Have no idea for really good alternatives, want to have them only be available there, that's why I find a user message as not so good way. As alternative can I use this:
```cpp
CEventLog::GetInstance().AddWithNotification(EventPtr(new CNotificationEvent(heading, text, EventLevel::Error)));
```
who bring:
![bildschirmfoto vom 2017-04-27 02-04-08](https://cloud.githubusercontent.com/assets/6879739/25462453/609cca4e-2aef-11e7-9b29-e052f6104ecc.png)

Here how it was with them here:
![bildschirmfoto vom 2017-04-27 02-09-31](https://cloud.githubusercontent.com/assets/6879739/25462460/75b302cc-2aef-11e7-956a-17e4f0a05e69.png)

garbear commented:
because it's an ABI change (break?) can we bumb the add-on versions? Otherwise an add-on compiled against a newer kodi will fail on an older Kodi build

ksooo commented:
> can we bumb the add-on versions?

 yes, please.

AlwinEsch commented:
Hi @garbear @ksooo this request is to allow it in future, currently is the add-on base interface not checked for version changes, that's the main reason to add `ADDON_GetTypeVersion` function here.

After them is in brings a old add-on a load failure, but no crashes:
![bildschirmfoto vom 2017-04-27 00-46-35](https://cloud.githubusercontent.com/assets/6879739/25460322/1abb6028-2ae3-11e7-8f11-4713150a95b9.png)

The Version increase does not work for them, because the library is not loaded and his functions not available.

After them comes another request who bring the new version check way, as example see this https://github.com/FernetMenta/kodi-agile/blob/master/addons/kodi.global.main/addon.xml.in#L2. This is only the first step.

AlwinEsch committed:
[addons] add new instance version check function

This alow the check for correct addon versions without code changes during
change on add-on.

AlwinEsch commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -461,5 +400,31 @@ ADDON_STATUS CAddonDll::TransferSettings()
+                                           kodi::addon::GetTypeName(type),
+                                           kodiVersion,
+                                           addonVersion);
+    CEventLog::GetInstance().AddWithNotification(EventPtr(new CNotificationEvent(Name(), text, EventLevel::Error)));
```
@tamland have changed it to the `CEventLog::GetInstance().AddWithNotification` with:
![bildschirmfoto vom 2017-04-27 02-04-08](https://cloud.githubusercontent.com/assets/6879739/25462752/68eb6e6a-2af1-11e7-97c0-989079a57a5e.png)

This by start of screensaver with wrong version:
![bildschirmfoto vom 2017-04-27 02-24-33](https://cloud.githubusercontent.com/assets/6879739/25462756/72b6aea0-2af1-11e7-91bb-f45eac56a1ae.png)

ksooo commented:
Not sure I got this right. But I guess you know best what you are doing. Maybe you should at least tell the addon maintainers that merging the addon PR you opened yesterday leads to a no longer functioning addon (at least until you merged this PR), because it is ABI incompatible?

AlwinEsch commented:
The function call tables (e.g. PVR) are not changed, this change "C" exported functions from the base addon functions.

Goal on end is to have only a `ADDON_Create` and `ADDON_GetTypeVersion` as exported functions, everything else becomes then handled with function tables. This allow then in future to check a several years old addon is compatible.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
OK to bring in?

stefansaraev commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/libKODI_guilib.h:
```diff
@@ -170,10 +170,10 @@ typedef struct CB_GUILib
 
 /* min. ADDONGUI API version */
-#define KODI_GUILIB_MIN_API_VERSION "5.10.0"
+#define KODI_GUILIB_MIN_API_VERSION ADDON_GLOBAL_VERSION_GUI
```
```
# cat /usr/share/kodi/addons/kodi.guilib/addon.xml 
<?xml version="1.0" encoding="UTF-8"?>
<addon id="kodi.guilib" version="#define KODI_GUILIB_API_VERSION ADDON_GLOBAL_VERSION_GUI" provider-name="Team-Kodi">
  <backwards-compatibility abi="#define KODI_GUILIB_MIN_API_VERSION ADDON_GLOBAL_VERSION_GUI"/>
  <requires>
    <import addon="xbmc.core" version="0.1.0"/>
  </requires>
</addon>
```

any addon that depends on guilib is probably broken now. 
https://github.com/xbmc/xbmc/blob/master/cmake/scripts/common/Macros.cmake#L632-L634 should also be changed
